File: 2023/python/day5.py
#!/usr/bin/env python3

from aoc import *


def part1(input: str):
    blocks = iter(input.split("\n\n"))
    seeds = comb(next(blocks), r"\d+", int, one_line=True)
    for entry in blocks:
        lines = iter(entry.splitlines())
        toks = next(lines).split()
        (f, t) = toks[0].split("-to-")
        mapping = []
        for line in lines:
            toks = [int(i) for i in line.split()]
            mapping.append((toks[1], toks[1] + toks[2] - 1, toks[0] - toks[1]))
        new_indexes = []
        for s in seeds:
            for m in mapping:
                if m[0] <= s <= m[1]:
                    new_indexes.append(s + m[2])
                    break
            else:
                new_indexes.append(s)
        seeds = new_indexes
    return min(seeds)


def part2(input: str):
    blocks = iter(input.split("\n\n"))
    ranges = iter(comb(next(blocks), r"\d+", int, one_line=True))
    seeds = []
    while True:
        try:
            start = next(ranges)
            num = next(ranges)
            seeds.append((start, start + num - 1))
        except StopIteration:
            break
    for entry in blocks:
        lines = iter(entry.splitlines())
        toks = next(lines).split()
        (f, t) = toks[0].split("-to-")
        mapping = []
        for line in lines:
            toks = [int(i) for i in line.split()]
            mapping.append((toks[1], toks[1] + toks[2] - 1, toks[0] - toks[1]))
        mapping.sort()
        new_ranges = []
        batch = iter(seeds)
        curr = next(batch)
        try:
            while curr:
                for m in mapping:
                    if m[0] <= curr[0] <= m[1]:
                        if curr[1] <= m[1]:
                            new_ranges.append((curr[0] + m[2], curr[1] + m[2]))
                            curr = next(batch)
                            break
                        else:
                            new_ranges.append((curr[0] + m[2], m[1] + m[2]))
                            curr = (m[1] + 1, curr[1])
                else:
                    new_ranges.append(curr)
                    curr = next(batch)
        except StopIteration:
            pass
        seeds = new_ranges
    return min(seeds)[0]


if __name__ == "__main__":
    main()

File: 2023/src/day5.rs
use std::{cmp::min, collections::VecDeque};

use aoc::*;
use nom::character::complete;

#[aoc_generator(day5)]
pub fn parse(input: &str) -> (Vec<i64>, Vec<Vec<SeedMap>>) {
    let mut groups = input.split("\n\n");
    let seeds = groups.next().map(comb(complete::i64)).unwrap();
    let transitions = groups
        .map(comb(complete::i64))
        .map(|t| {
            let mut sm: Vec<SeedMap> = t
                .chunks(3)
                .map(|x| SeedMap {
                    start: x[1],
                    end: x[1] + x[2] - 1,
                    diff: x[0] - x[1],
                })
                .collect();
            sm.sort();
            sm
        })
        .collect();
    (seeds, transitions)
}

#[derive(Debug, Eq, Ord, PartialEq, PartialOrd)]
pub struct SeedMap {
    start: i64,
    end: i64,
    diff: i64,
}

#[aoc(day5, part1)]
pub fn part1((seeds, transitions): &(Vec<i64>, Vec<Vec<SeedMap>>)) -> i64 {
    seeds
        .iter()
        .map(|s| {
            transitions.iter().fold(*s, |seed, t| {
                for sm in t {
                    if seed >= sm.start && seed <= sm.end {
                        return seed + sm.diff;
                    }
                }
                seed
            })
        })
        .min()
        .unwrap()
}

#[aoc(day5, part2)]
pub fn part2((seeds, transitions): &(Vec<i64>, Vec<Vec<SeedMap>>)) -> i64 {
    let mut smallest = i64::MAX;
    let mut seeds: VecDeque<(usize, i64, i64)> =
        seeds.chunks(2).map(|r| (0, r[0], r[0] + r[1])).collect();
    while let Some((g, mut start, mut end)) = seeds.pop_back() {
        for (i, seed_maps) in transitions.iter().enumerate().skip(g) {
            for t in seed_maps {
                if start >= t.start && start <= t.end {
                    if end <= t.end {
                        start += t.diff;
                        end += t.diff;
                        break;
                    }
                    seeds.push_back((i + 1, start + t.diff, t.end + t.diff));
                    start = t.end + 1;
                }
            }
        }
        smallest = min(smallest, start);
    }
    smallest
}
